testing/trino-product-tests/src/main/java/io/trino/tests/product/utils/DeltaQueryExecutors.java

```diff
@@ -0,0 +1,164 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.tests.product.utils;
+
+import com.google.common.collect.ImmutableList;
+import dev.failsafe.Failsafe;
+import dev.failsafe.RetryPolicy;
+import io.airlift.log.Logger;
+import io.airlift.resolver.ArtifactResolver;
+import io.airlift.resolver.DefaultArtifact;
+import io.trino.tempto.context.TestContext;
+import io.trino.tempto.query.JdbcConnectionsPool;
+import io.trino.tempto.query.JdbcConnectivityParamsState;
+import io.trino.tempto.query.JdbcQueryExecutor;
+import io.trino.tempto.query.QueryExecutor;
+import org.sonatype.aether.artifact.Artifact;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import java.lang.reflect.Field;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.sql.Connection;
+import java.sql.Driver;
+import java.sql.SQLException;
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.airlift.resolver.ArtifactResolver.MAVEN_CENTRAL_URI;
+import static io.airlift.resolver.ArtifactResolver.USER_LOCAL_REPO;
+import static io.trino.tempto.context.ThreadLocalTestContextHolder.testContext;
+import static java.lang.ClassLoader.getPlatformClassLoader;
+
+public final class DeltaQueryExecutors
+{
+    private static final Logger log = Logger.get(DeltaQueryExecutors.class);
+
+    @GuardedBy("DRIVERS")
+    private static final Map<String, Driver> DRIVERS = new HashMap<>();
+
+    private static final RetryPolicy<Driver> loadDatabaseDriverRetryPolicy = RetryPolicy.<Driver>builder()
+            .withMaxRetries(30)
+            .withDelay(Duration.ofSeconds(10))
+            .onRetry(event -> log.warn(event.getLastException(), "Download failed on attempt %d, will retry.", event.getAttemptCount()))
+            .build();
+
+    private static final DeltaConnectionsPool deltaConnectionPool = new DeltaConnectionsPool();
+
+    private DeltaQueryExecutors() {}
+
+    public static JdbcQueryExecutor configureConnectionsPool(TestContext testContext)
+    {
+        JdbcQueryExecutor jdbcQueryExecutor = (JdbcQueryExecutor) testContext().getDependency(QueryExecutor.class, "delta");
+        try {
+            Field field = jdbcQueryExecutor.getClass().getDeclaredField("jdbcConnectionsPool");
+            field.setAccessible(true);
+            field.set(jdbcQueryExecutor, deltaConnectionPool);
+            testContext.registerCloseCallback(context -> deltaConnectionPool.close());
+            return jdbcQueryExecutor;
```

Comment:
Its all our code. We shouldn't need any reflective access

Reply:
```java

    public static QueryExecutor createDeltaQueryExecutor(TestContext testContext)
    {
        JdbcConnectivityParamsState connectivityParams = testContext.getDependency(JdbcConnectivityParamsState.class, "delta");
        Driver driver = getDriverFromMaven(connectivityParams.driverClass);

        return new JdbcQueryExecutor(
                connectivityParams,
                new JdbcConnectionsPool()
                {
                    @Override
                    protected Connection createConnection(JdbcConnectivityParamsState jdbcParamsState)
                            throws SQLException
                    {
                        Properties properties = new Properties();
                        properties.put("user", jdbcParamsState.user);
                        properties.put("password", jdbcParamsState.password);
                        return driver.connect(jdbcParamsState.url, properties);
                    }
                },
                testContext); // registers close callback
    }

    private static Driver getDriverFromMaven(String className)
    {
        synchronized (DRIVERS) {
            return DRIVERS.computeIfAbsent(className, DeltaQueryExecutors::loadDriverFromMaven);
        }
    }

    private static Driver loadDriverFromMaven(String className)
    {
        return Failsafe.with(loadDatabaseDriverRetryPolicy)
                .get(() -> doLoadDriverFromMaven(className));
    }

    private static Driver doLoadDriverFromMaven(String driverClassName)
    {
        // TODO Add support for maven coordinate in tempto
        ArtifactResolver resolver = new ArtifactResolver(USER_LOCAL_REPO, ImmutableList.of(MAVEN_CENTRAL_URI));
        List<URL> classPath = resolver.resolveArtifacts(getArtifact(driverClassName)).stream()
...
```

Reply:
The code doesn't work when tests depend on session likes `testDatabricksVacuumRemoveChangeDataFeedFiles `

Reply:
Good point. 
We can implement `io.trino.tempto.TemptoPlugin#getTestModules` in `io.trino.tests.product.PrestoTemptoPlugin` and inject something that will hold the query executor so that `onDelta()` can get-or-create.